In [7]:
import numpy as np
import time
from sklearn.linear_model import LinearRegression

In [8]:
def compute_mse_vectorized(w,X,Y):
    res = Y - np.dot(X,w)
    totalError = np.dot(res.T,res)
    return totalError / float(len(Y))

In [9]:
def step_gradient_vectorized(w_current,X,Y,learningRate):
    res = Y - np.dot(X,w_current)
    gradient = np.multiply(res,X)
    gradient = np.sum(gradient,axis=0)
    gradient = gradient[:,np.newaxis]
    new_w = w_current + 2 * learningRate * gradient
    return [new_w,gradient]

In [10]:
def gradient_descent_runner_vectorized(starting_w, X,Y, learning_rate, epsilon):
    w = starting_w
    grad = np.array([np.inf])
    while (np.linalg.norm(grad)>=epsilon):
        w,grad = step_gradient_vectorized(w, X, Y, learning_rate)
    return w

In [11]:
points = np.genfromtxt("sample_treino.csv", delimiter=",")
points = np.c_[np.ones(len(points)),points]
X = points[:,0:-1]
Y = points[:,-1][:,np.newaxis]

init_w = np.zeros((len(X[0]),1))
learning_rate = 0.00003
epsilon = 0.01
print("Iniciando gradiente descendente com todos os parâmetros iniciais 0 e erro = {0}".format(compute_mse_vectorized(init_w, X,Y)[0][0]))
print("Running...")
tic = time.time()
w = gradient_descent_runner_vectorized(init_w, X,Y, learning_rate, epsilon)
toc = time.time()
print("Gradiente descendente convergiu com w0 = {0}, w1 = {1}, w2 = {2}, w3 = {3}, w4 = {4}, w5 = {5}, error = {6}".format(w[0][0], w[1][0], w[2][0], w[3][0], w[4][0], w[5][0], compute_mse_vectorized(w,X,Y)[0][0]))
print("Versão vetorizada rodou em: " + str(1000*(toc-tic)) + " ms")

Iniciando gradiente descendente com todos os parâmetros iniciais a 0 e erro = [[ 54.47995386]]
Running...
Gradiente descendente convergiu com w0 = 1.7214319757862273, w1 = 0.10333313833600129, w2 = 0.04718170808359163, w3 = 0.16406125198726337, w4 = 0.38200508559551033, w5 = 0.02047849242976369, error = 0.41133944383813986
Versão vetorizada rodou em: 3037.3594760894775 ms


In [12]:
reg = LinearRegression()
reg.fit(X,Y)
print(reg.coef_)

[[ 0.          0.10304143  0.0464367   0.16409834  0.38117843  0.02027816]]


É possível observar que os coeficientes obtidos usando a regressão linear do scikit-learn e os obtidos com o gradiente descendente são muito próximos.
A maior diferença está no coeficiente w0 que é igual a 0 pelo scikit-learn, e 1.72 pelo gradiente descendente, todos os outros coeficiente possuem diferença menor que 0.001.